flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/DefaultVertexParallelismDecider.java

```diff
@@ -81,52 +88,36 @@ private int calculateParallelism(List<BlockingResultInfo> consumedResults) {
-                            + "'. "
-                            + "Parallelism will be set to {}.",
-                    maxParallelism);
-
-            return maxParallelism;
-        } else if (broadcastBytes == dataVolumePerTask) {
-            return minParallelism;
-        } else {
-            int parallelism =
-                    (int)
-                            Math.ceil(
-                                    (double) nonBroadcastBytes
-                                            / (dataVolumePerTask - broadcastBytes));
-            parallelism = Math.max(parallelism, minParallelism);
-            parallelism = Math.min(parallelism, maxParallelism);
-            return parallelism;
-        }
-    }
-
-    /** The factory to instantiate {@link DefaultVertexParallelismDecider}. */
-    public static class Factory implements VertexParallelismDecider.Factory {
+        long expectedMaxBroadcastBytes =
+                (long) Math.ceil((dataVolumePerTask * CAP_RATIO_OF_BROADCAST));
 
-        private final Configuration configuration;
+        if (broadcastBytes > expectedMaxBroadcastBytes) {
+            LOG.info(
+                    "The number of broadcast bytes: {} is larger than the expected maximum value: {} ('{}' * {})."
+                            + " Use the expected maximum value as the number of broadcast bytes to decide the parallelism.",
+                    broadcastBytes,
+                    expectedMaxBroadcastBytes,
+                    JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_DATA_VOLUME_PER_TASK.key(),
+                    CAP_RATIO_OF_BROADCAST);
 
-        public Factory(Configuration configuration) {
-            this.configuration = configuration;
+            broadcastBytes = expectedMaxBroadcastBytes;
         }
 
-        @Override
-        public VertexParallelismDecider create() {
-            return new DefaultVertexParallelismDecider(
-                    configuration.getInteger(
-                            JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_MAX_PARALLELISM),
-                    configuration.getInteger(
-                            JobManagerOptions.ADAPTIVE_BACH_SCHEDULER_MIN_PARALLELISM),
-                    configuration.get(
-                            JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_DATA_VOLUME_PER_TASK),
-                    configuration.get(
-                            JobManagerOptions.ADAPTIVE_BATCH_SCHEDULER_DEFAULT_SOURCE_PARALLELISM));
-        }
+        int parallelism =
+                (int) Math.ceil((double) nonBroadcastBytes / (dataVolumePerTask - broadcastBytes));
+        parallelism = Math.max(parallelism, minParallelism);
+        parallelism = Math.min(parallelism, maxParallelism);
+        return parallelism;
+    }
+
+    public static DefaultVertexParallelismDecider from(Configuration configuration) {
+        return new DefaultVertexParallelismDecider(
```

Comment:
I think we need to do some checks about the params, like minParallelism> 0, minParallelism < maxParallelism, dataVolumePerTask > 0, defaultSourceParallelism > 0.

Reply:
Yes, you are right. I will add the checks in ctor. The  `dataVolumePerTask` is `MemorySize` type, I will check it's not null.